### PR TITLE
HOTT-4135: Backend Admin APIs

### PR DESCRIPTION
**Pull request by amberstarlight**
### Jira link

[HOTT-4135](https://transformuk.atlassian.net/browse/HOTT-4135)

### What?

I have added/removed/altered:

- Adds backend admin API services.
- Updates pre-commit hooks.
- Adds db schema migration for adding a read-only postgres user.

### Why?

I am doing this because:

- In order to ensure that none of our user-facing services have read-write to the database, and therefore no attack surface, we need an 'admin' API for the backend, which is the sole read-writer to the database.
- This PR adds 'admin' backend API services. The admin tool will be updated to point to these backends.
- We need a database user that has read-only (`SELECT`) on all tables that we can use in the APIs.

### CODE REVIEW

File: db/migrate/20231030151920_read_only_user.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+user = 'tariff_read'
+db = Sequel::Model.db
+
+Sequel.migration do
+  up do
+    # Create a user with SELECT on all tables, including tables created after this migration
+    db.run(sprintf('CREATE USER %s', user))
```
Comment (willfish) [inline db/migrate/20231030151920_read_only_user.rb:9]:
```suggestion
    db.run("CREATE USER #{user}")
```

File: terraform/admin_uk.tf [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  task_role_policy_arns = [
+    aws_iam_policy.s3.arn,
+    aws_iam_policy.task_role_kms_keys.arn
+  ]
+
+  execution_role_policy_arns = [
+    aws_iam_policy.secrets.arn
+  ]
+
+  init_container            = true
```
Comment (willfish) [inline terraform/admin_uk.tf:35]: I think we need to make sure that migrations aren't run on this app?

File: terraform/admin_xi.tf [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+      {
+        name  = "CDS"
+        value = "true"
+      },
+      {
+        name  = "GOVUK_APP_DOMAIN"
+        value = "tariff-xi-admin-${var.environment}.apps.internal" # This is necessary for a GOVUK gem we're not using
+      },
+      {
+        name  = "NEW_RELIC_APP_NAME"
```
Comment (willfish) [inline terraform/admin_xi.tf:55]: Curious! I thought I'd murdered these

File: db/migrate/20231213114821_read_only_user.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
```
Comment (jebw) [inline db/migrate/20231213114821_read_only_user.rb:1]: I'd expect to see corresponding changes in `structure.sql`

File: .gitignore [hunk 5]
```diff
@@ -40,5 +40,9 @@ searches_without_match_or_synonym.csv
 # Ignore all json files in the cds dump directory
 /data/cds_dumps/*.json
 
+# Ignore all database dumps
+*.psql
```
Comment (jebw) [inline .gitignore:44]: we dump gzipped files so probably `*.psql.gz` and `*.sql.gz` as well

File: .gitignore [hunk 6]
```diff
@@ -40,5 +40,9 @@ searches_without_match_or_synonym.csv
...
+*.sql
```
Comment (jebw) [inline .gitignore:45]: need an exception for `db/structure.sql`

**Review by jebw — COMMENTED** [review]
Looks good, couple of comments around SQL files

File: terraform/admin_uk.tf [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+      {
+        name  = "SLACK_USERNAME"
+        value = "UK Backend API ${title(var.environment)}"
+      },
+      {
+        name  = "TARIFF_FROM_EMAIL"
+        value = "Tariff UK [${title(var.environment)}] <${local.no_reply}>"
+      },
+      {
+        name  = "VCAP_APPLICATION"
```
Comment (willfish) [inline terraform/admin_uk.tf:60]: Remove me

File: terraform/worker_uk.tf [hunk 8]
```diff
@@ -74,11 +74,15 @@ module "worker_uk" {
...
+    local.backend_common_secrets,
+    local.backend_uk_common_secrets,
+    local.backend_uk_worker_secrets,
     [
-      local.backend_common_secrets,
-      local.backend_uk_common_secrets,
-      local.backend_uk_worker_secrets,
+      {
+        name      = "DATABASE_URL"
+        valueFrom = data.aws_secretsmanager_secret.database_readonly_connection_string.arn
```
Comment (willfish) [inline terraform/worker_uk.tf:84]: These want the read/write

File: terraform/worker_xi.tf [hunk 9]
```diff
@@ -74,11 +74,15 @@ module "worker_xi" {
...
+    local.backend_common_secrets,
+    local.backend_xi_common_secrets,
+    local.backend_xi_worker_secrets,
     [
-      local.backend_common_secrets,
-      local.backend_xi_common_secrets,
-      local.backend_xi_worker_secrets,
+      {
+        name      = "DATABASE_URL"
+        valueFrom = data.aws_secretsmanager_secret.database_readonly_connection_string.arn
```
Comment (willfish) [inline terraform/worker_xi.tf:84]: Same again

File: terraform/admin_uk.tf [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        name  = "GOVUK_APP_DOMAIN"
+        value = "tariff-uk-admin-${var.environment}.apps.internal" # This is necessary for a GOVUK gem we're not using
+      },
+      {
+        name  = "SERVICE"
+        value = "uk"
+      },
+      {
+        name  = "SLACK_USERNAME"
+        value = "UK Backend API ${title(var.environment)}"
```
Comment (willfish) [inline terraform/admin_uk.tf:53]:
```suggestion
        value = "UK Backend Admin API ${title(var.environment)}"
```

File: terraform/admin_xi.tf [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        name  = "GOVUK_APP_DOMAIN"
+        value = "tariff-xi-admin-${var.environment}.apps.internal" # This is necessary for a GOVUK gem we're not using
+      },
+      {
+        name  = "SERVICE"
+        value = "xi"
+      },
+      {
+        name  = "SLACK_USERNAME"
+        value = "XI Backend API ${title(var.environment)}"
```
Comment (willfish) [inline terraform/admin_xi.tf:53]:
```suggestion
        value = "XI Backend Admin API ${title(var.environment)}"
```

**Review by willfish — DISMISSED** [review]

**Comment by willfish** [comment]
Let's get this merged and manually kick off the regression suite <3

**Review by willfish — APPROVED** [review]

**Review by willfish — DISMISSED** [review]

**Comment by willfish** [comment]
Done some tests:
1. Validated that the admin CRUD operations are fine in dev and that
2. All of the read only apis are returning healthily

Also, these smoketest failures relate to a broken 7.1 frontend and CSRF tokens not matching up

**Pull request merged**